### ISSUE TITLE
[Enhancement] implement patch to add FULLSYNC capability (Time machine backups)

### ISSUE THREAD

**Issue by nidstigator** [wongsyrone/lede-1#34]
Hi there,

This is a long shot, but could you by any chance look at implementing this PR in your SAMBA package when you get some time? It would be great to get rid of AFP, since I only use it for Time Machine backups for my macbook.

Thanks!

EDIT: forgot to add link to the mentioned PR:

https://github.com/samba-team/samba/pull/64

**Comment by wongsyrone** [comment]
Why not waiting for new release?

**Comment by nidstigator** [comment]
indications are weak that this will be accepted anytime soon and go into a release.
Also, judging by the thread on that PR, everyone who used it found that it's working perfectly.

Please let me know if you think about including it! I'll be happy to test!

**Comment by inspector-ambitious** [comment]
@nidstigator I added the patch in this commit, just compiling right now and testing soon

https://github.com/yhwh/lede-1/commit/ed36abb2b73cf315f494401e77c1d7f5d5645ca5

**Comment by wongsyrone** [comment]
Awesome. I don't have such device to test.

**Comment by inspector-ambitious** [comment]
I will do some test next week to try to turn my WRT1900ACS into a time capsule.

**Comment by inspector-ambitious** [comment]
added a patch for python https://github.com/wongsyrone/packages-1/pull/9

**Comment by inspector-ambitious** [comment]
also need that patch to compile samba https://github.com/commodo/packages/commit/0bb78f682519efafeb69cae251bf1d158effbd51

**Comment by nidstigator** [comment]
@yhwh Thank you!! compiled successfully. In the process of setting up timemachine now :) Will report back here.

**Comment by nidstigator** [comment]
@yhwh So I spent all day trying to set up TimeMachine. I cannot get avahi to advertise the share no matter what I try.

I have tried the following: 

```
<?xml version="1.0" standalone='no'?>
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">
<service-group>
 <name replace-wildcards="yes">%h</name>
 <service>
   <type>_smb._tcp</type>
   <port>445</port>
 </service>
 <service>
   <type>_device-info._tcp</type>
   <port>0</port>
   <txt-record>model=RackMac</txt-record>
 </service>
 <service>
   <type>_adisk._tcp</type>
   <txt-record>sys=waMa=0,adVF=0x100</txt-record>
   <txt-record>dk0=adVN=TimeMachine,adVF=0x82</txt-record>
 </service>
</service-group>

```
Which seems to be working with everyone, but I get connection failed. 
The share is working fine through SMB, can access it and write to it (no permission issues).

Were you able to set it up? If so, please let me know how you did it.

Thanks

**Comment by inspector-ambitious** [comment]
@nidstigator haven't tried yet I have family who came over for the week end. Will give it a go on Tuesday I guess. Would you mind sharing your smb.conf as well ?
Also you may want to have a look at the avahi configuration from https://wiki.openwrt.org/doc/howto/timemachine
It seems the share name should match adVN=TimeMachine, haven't investigated yet the dk0 adVF parameters

**Comment by nidstigator** [comment]
@yhwh Yes, I have got TimeMachine to work through AFP before using that very same article you've linked. 

However, using the SMB compiled from here, the Bonjour entry for TimeMachine shows up, but doesn't request any credentials and fails to connect immediately. 

Yes, I also made sure the VN is TimeMachine. adVF should be 0x82 per Apple specs page. 

Anyway, here is my smb.conf:

```
[global]
        netbios name = Home-Network
        display charset = UTF-8
        interfaces = 192.168.1.1/255.255.255.0 \
                      192.168.4.1/255.255.255.0 \
                      192.168.5.1/255.255.255.0 \
                      192.168.6.1/255.255.255.0 \
        server string = Home network volumes
        unix charset = UTF-8
        workgroup = Milkyway
        local master = yes
        browseable = yes
        deadtime = 30
        domain master = yes
        encrypt passwords = yes
        enable core files = no
        invalid users = root
        load printers = no
        map to guest = Bad User
        passdb backend = tdbsam
        preferred master = yes
        security = user
        syslog = 2
        use sendfile = no
        bind interfaces only = yes
        socket options = TCP_NODELAY IPTOS_LOWDELAY SO_RCVBUF=65536 SO_SNDBUF=65
        name resolve order = wins lmhosts hosts bcast
        os level = 99
        durable handles = yes
        kernel oplocks = no
        kernel share modes = no
        posix locking = no
        fruit:advertise_fullsync = true
        wins support = yes

[WindowsBackup]
        path = /mnt/WindowsBackup
        writable = yes
        create mask = 0664

[TimeMachine]
        path = /mnt/Shared
        writable = yes
        create mask = 0640

[Torrents]
        path = /mnt/Torrenting
        writable = yes
        create mask = 0664
```


samba4 config file:

```
config samba
        option homes '0'
        option name 'Home-Network'
        option description 'Home network volumes'
        option workgroup 'Milkyway'
        option enabled                '1'

config sambashare
        option read_only 'no'
        option guest_ok 'yes'
        option browsable 'yes'
        option writable 'yes'
        option read_list 'guest nobody'
        option write_list 'kiki nidstigator'
        option name 'WindowsBackup'
        option path '/mnt/WindowsBackup'
        option create_mask '0664'
        option dir_mask '0775'

config sambashare
        option name 'TimeMachine'
        option path '/mnt/Shared'
        option browseable 'yes'
        option guest_ok 'no'
        option vfs_objects 'catia fruit streams_xattr'
        option fruit:aapl 'yes'
        option read_only 'no'
        option writable 'yes'
        option write_list 'tmuser'
        option create_mask '0640'
        option dir_mask '2755'

config sambashare
        option name 'Torrents'
        option path '/mnt/Torrenting'
        option read_only 'no'
        option guest_ok 'yes'
        option browsable 'yes'
        option read_list 'guest nobody'
        option write_list 'kiki nidstigator'
        option create_mask '0664'
        option dir_mask '0775'
        option writable 'yes'
```

**Comment by nidstigator** [comment]
And here is the apple specs for TimeMachine over SMB:

https://developer.apple.com/library/content/releasenotes/NetworkingInternetWeb/Time_Machine_SMB_Spec/

**Comment by nidstigator** [comment]
@yhwh FWIW, what's most confusing when setting it up is the syntax differences between smb.conf and the samba4 config file in /etc/config. (i.e spaces vs _)

Anyway, I hope it works out when you try it out!

**Comment by inspector-ambitious** [comment]
looking at it @nidstigator the init script creates a  ```/var/etc/smb.conf``` don't have time to sort this out tonight but seems to be the one used by the server

**Comment by nidstigator** [comment]
@yhwh This is created according to settings found in /etc/config/samba4.

**Comment by nidstigator** [comment]
@yhwh I found out why samba is ignoring my configuration when adding the fullsync specific attributes for the time machine share; it's because the init.d script ignores it. 

I've added the following in the init_config function in the init.d script:

```
        config_get fruit_aapl $1 fruit_aapl
        config_get vfs_objects $1 vfs_objects

        [ -n "$fruit_aapl" ] && echo -e "\tfruit:aapl = $fruit_aapl" >> /var/etc/smb.conf
        [ -n "$vfs_objects" ] && echo -e "\tvfs objects = $vfs_objects" >> /var/etc/smb.conf
```

so now smb doesn't ignore these params when defined in /etc/config/samba4.


The resulting smb.conf is:

```
[global]
        netbios name = Home-Network
        display charset = UTF-8
        interfaces = 192.168.1.1/255.255.255.0 \
                      192.168.4.1/255.255.255.0 \
                      192.168.5.1/255.255.255.0 \
                      192.168.6.1/255.255.255.0 \
        server string = Home network volumes
        unix charset = UTF-8
        workgroup = Milkyway
        browseable = yes
        deadtime = 30
        encrypt passwords = yes
        enable core files = no
        invalid users = root
        load printers = no
        map to guest = Bad User
        passdb backend = tdbsam
        security = user
        syslog = 2
        bind interfaces only = yes
        socket options = TCP_NODELAY IPTOS_LOWDELAY SO_RCVBUF=65536 SO_SNDBUF=65
        durable handles = yes
        kernel oplocks = no
        kernel share modes = no
        posix locking = no
        os level = 20
        preferred master = yes
        local master = yes
        domain master = yes
        guest ok = yes
        guest account = nobody
        fruit:advertise_fullsync = true

[WindowsBackup]
        path = /mnt/WindowsBackup
        writable = yes
        create mask = 0664
        write list = kiki nidstigator
        directory mask = 0775

[TM_Backup_Network]
        path = /mnt/Shared
        writable = yes
        create mask = 0664
        fruit:aapl = yes
        vfs objects = catia fruit streams_xattr
        directory mask = 0775

[Torrents]
        path = /mnt/Torrenting
        writable = yes
        create mask = 0664
        write list = kiki nidstigator
        directory mask = 0775
```


However, when trying to mount the share now, I get: 

smb_mount: mount failed to home-network/TM_Backup_Network, syserr = Unknown error: -1073741275



I am lost.

**Comment by inspector-ambitious** [comment]
@nidstigator  So I just did some tries and bypassed the init options in a monkey patch manner if you get my latest changes 0951c54c8bdbd0771ece1c678ea53e8060ad0e50
you should have catia fruit streams_xattr compiled statically into it.
At this point I'm able to see it with avahi daemon configured properly.
But I get this error message in TimeMachine after authentification "The selected network backup disk does not support the required capabilities."
Out of curiosity what is your version of MacOS ?
Feel free to send a pr to my repo with the changes for the init file it's way cleaner than my monkey patch btw

**Comment by inspector-ambitious** [comment]
I fear I may have to compile a more recent version maybe 4.4.7 or 4.4.8
Edit: recompiling right now with 4.4.8 8c029a52cfd21a7f7f52d6b53ab2dbb6ef9a828f

**Comment by inspector-ambitious** [comment]
if it doesn't work going to give a shot at 4.6/4.8 branches... such a pain...

**Comment by inspector-ambitious** [comment]
wow seems I got it working in fact the guy changed the option name in his last refactor. read the pr attentively in fact it's fruit:time machine = yes instead of fruit:advertise_fullsync = true

**Comment by inspector-ambitious** [comment]
going to test this over next week and cleanup my mess to have a 4.5.8 clean install

**Comment by inspector-ambitious** [comment]
Sharing my final smb.conf
```
[global]
    workgroup = SAMBA
    security = user
    passdb backend = tdbsam
    durable handles = yes
    kernel oplocks = no
    kernel share modes = no
    posix locking = no
    socket options = TCP_NODELAY IPTOS_LOWDELAY SO_RCVBUF=524288 SO_SNDBUF=524288
    fruit:time machine = yes
    hosts allow = 192.168.1. 127.
    interfaces = 127.0.0.1/8 lo 192.168.1.1/24 br-lan
    invalid users = root
    encrypt passwords = yes

[TimeMachine]
    path = /mnt/TimeMachine
    browseable = yes
    fruit:time machine = yes
    vfs objects = catia fruit streams_xattr
    fruit:aapl = yes
    valid users = tmuser
    read only = No
    inherit aclis = yes
```

**Comment by inspector-ambitious** [comment]
@nidstigator if you want to give it a go with samba 4.5.8 
7e95f242cdab3b9db19773bc98703dd9fed7f18f
Use my smb.conf directly (I disabled /etc/config/samba4)
Sharing my avahi-daemon.conf
```
[server]
#host-name=foo
#domain-name=local
#browse-domains=0pointer.de, zeroconf.org
use-ipv4=yes
use-ipv6=yes
#allow-interfaces=eth0
#deny-interfaces=eth1
#check-response-ttl=no
#use-iff-running=no
enable-dbus=no
#disallow-other-stacks=no
#allow-point-to-point=no
#cache-entries-max=4096
#clients-max=4096
#objects-per-client-max=1024
#entries-per-entry-group-max=32
ratelimit-interval-usec=1000000
ratelimit-burst=1000

[wide-area]
enable-wide-area=yes

[publish]
#disable-publishing=no
#disable-user-service-publishing=no
#add-service-cookie=no
#publish-addresses=yes
publish-hinfo=no
publish-workstation=no
#publish-domain=yes
#publish-dns-servers=192.168.50.1, 192.168.50.2
#publish-resolv-conf-dns-servers=yes
#publish-aaaa-on-ipv4=yes
#publish-a-on-ipv6=no

[reflector]
#enable-reflector=no
#reflect-ipv=no

[rlimits]
#rlimit-as=
rlimit-core=0
rlimit-data=4194304
rlimit-fsize=0
rlimit-nofile=768
rlimit-stack=4194304
rlimit-nproc=3
```

and my smbd.service

```
root@Mir:~# cat /etc/avahi/services/smbd.service
<?xml version="1.0" standalone='no'?>
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">
<service-group>
 <name replace-wildcards="yes">%h</name>
 <service>
   <type>_smb._tcp</type>
   <port>445</port>
 </service>
 <service>
   <type>_device-info._tcp</type>
   <port>0</port>
   <txt-record>model=TimeCapsule</txt-record>
 </service>
 <service>
   <type>_adisk._tcp</type>
   <port>0</port>
   <txt-record>dk0=adVN=TimeMachine,adVF=0x82</txt-record>
 </service>
</service-group>
```

it works for me so far (the first backup is a bit long, only 35G in 5h and I'm on 5Ghz at 1Gbps)
Also I had to put  fruit:time machine = yes in both [global] and [TimeMachine], so I recommend to stick exactly to my config. If you feel like cleaning the samba4.init feel free I'll merge it (right now I'm bypassing everything). Was a damn journey to make this work man !

**Comment by nidstigator** [comment]
Hey man, thanks for all of this. I will compile 4.5.8 and start from scratch and report back later.

Thanks for all the effort!

**Comment by inspector-ambitious** [comment]
needed it badly, I may update to samba 4.6.latest

**Comment by nidstigator** [comment]
@yhwh Hey, got 4.5.8 installed. Avahi advertises the volume and it is TM capable. My final smb.conf is the following:


```
[global]
        netbios name = Home-Network
        interfaces = br-lan USVPN UKVPN virtualinterface
        server string = Home network volumes
        workgroup = Milkyway
        deadtime = 30
        encrypt passwords = yes
        map to guest = Bad User
        passdb backend = tdbsam
        socket options = TCP_NODELAY IPTOS_LOWDELAY SO_RCVBUF=65536 SO_SNDBUF=65
        durable handles = yes
        kernel oplocks = no
        kernel share modes = no
        posix locking = no
        preferred master = yes
        local master = yes
        domain master = yes
        guest ok = yes
        invalid users = root
        valid users = @sambausers, tmuser


[TM_Backup_Network]
        path = /mnt/Shared
        create mask = 0774
        fruit:aapl = yes
        vfs objects = catia fruit streams_xattr
        directory mask = 0774
        fruit:time machine = yes
        read only = no
        inherit acls = yes
        guest ok = no
        write list = tmuser
```


I have same Avahi service file as you. 

Regarding cleaning up the init script, I think editing smb.conf directly makes more sense.

Thanks for all the help and the experimentation! Please let me know if you bump the SMB version I could try and rebuild and check again.

**Comment by inspector-ambitious** [comment]
It's possible that I may add a vfruit:max_disk_space option for the share to limit the disk usage for the timemachine share. I was looking at vfs_fruit.c seems doable.
like 
```
vfruit:max_disk_space = 1024  # limit to 1GB
```

**Comment by inspector-ambitious** [comment]
in fact just created partitions on the external hard drive this way I can backup my macbookpro on one share and my wife macbook air on another one (just need to add an extra txt-record entry in the avahi service). This is damn neat :)

**Comment by nidstigator** [comment]
Yes, that's what I do. Separate partitions. 

I tried bumping up version to 4.6.latest but most patches break.  So I'll hold off upgrading to 4.6 for now.

**Comment by inspector-ambitious** [comment]
Not going to touch this anymore seems to work. Did some experiments and kept an eye on the logs.

**Comment by nidstigator** [comment]
Yes! Closing this issue :) 
Thanks again for the help!

**Comment by nidstigator** [comment]
@yhwh Hey, have you noticed that available memory drops to 1-5% when something intensive such as backing up Windows/Time Machine to an SMB share?

**Comment by nidstigator** [comment]
@yhwh Sometimes this leads to a complete router crash for me and reboots the router. My router is WRT1900ACS V2 and I am running one of LEDE stable builds. (Kernel 4.4.52).

**Comment by inspector-ambitious** [comment]
Just noticed a crash as well. Haven't checked the memory yet.
The annoying bit is that i had to repair hfs manually 
But I compiled https://github.com/dottedmag/openwrt.hfsprogs
and set fsck_enabled to 1. in /etc/config/mount
gonna check to see if we can reduce the memory munching and see what proc is responsable

**Comment by inspector-ambitious** [comment]
Did some changes to the global section from the original template, it may help save memory maybe

```
[global]
    workgroup = SAMBA
    security = user
    passdb backend = tdbsam
    durable handles = yes
    kernel oplocks = no
    kernel share modes = no
    posix locking = no
    fruit:time machine = yes
    hosts allow = 192.168.1. 127.
    interfaces = 127.0.0.1/8 lo 192.168.1.1/24 br-lan
    invalid users = root
    encrypt passwords = yes
    strict allocate = yes
    strict locking = no
    socket options = TCP_NODELAY IPTOS_LOWDELAY
    use sendfile = yes
    max xmit = 65535
    max connections = 8
    deadtime = 30
    dns proxy = no
    aio read size = 0
    aio write size = 0
```

**Comment by inspector-ambitious** [comment]
maybe some swap would help ?

**Comment by nidstigator** [comment]
I have always had 1GB (I know, too much) of swap available. Only 1% gets used during the memory craziness. 
Let me know how the new global variables work! :)

**Comment by inspector-ambitious** [comment]
still low mem available

**Comment by inspector-ambitious** [comment]
but crashed only once so far

**Comment by nidstigator** [comment]
There are vfs_fruit related mem leaks that were fixed in 4.5.x according to dev forums.
Are you still running 4.4?

**Comment by inspector-ambitious** [comment]
4.5.8

**Comment by nidstigator** [comment]
@yhwh just out of curiosity, what filesystem are you using on your timemachine drive?

I have recompiled 4.5.8 and reset up my router and running into issues now; I am using the exact config file you attached above and get the "The network backup disk does not support the required capabilities" when attempting to start backup!

**Comment by nidstigator** [comment]
@yhwh nvm, it was a permission issue. Sorry.

**Comment by inspector-ambitious** [comment]
Also if you want to put in place user quotas at the root of the share in case you want to use one partition.
/mnt/storage/Backups/.com.apple.TimeMachine.quota.plist with the following. This will create a 250GB per user limit

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>GlobalQuota</key>
	<integer>250000000000</integer>
</dict>
</plist>
```
I think I'm going to remove the limit on max connections and deadtime in the smb.conf

**Comment by nidstigator** [comment]
Thanks.
Still crashing here. We might just have to accept that this needs more power than the routers can provide.

**Comment by Andy2244** [comment]
Hi all, my samba4.8 package just got merged into openwrt master and i have some requests to get timemachine working. I will move the related vfs modules around so they are all together and available by default, so timemachine can work.

What options exist to announce timemachine on mac? I don't build nmbd/netbios by default anymore, since i use a WSD daemon for windows clients. So is avahi the only option for mac?
